packages/report-server/src/__tests__/reportBuilder/output/output.test.ts

```diff
@@ -0,0 +1,91 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
+ */
+
+import { buildOutput } from '../../../reportBuilder/output';
+import { MULTIPLE_TRANSFORMED_DATA } from './output.fixtures';
+
+describe('output', () => {
+  describe('matrix', () => {
+    const expectedData = {
+      columns: [
+        {
+          key: 'Laos',
+          title: 'Laos',
+        },
+        {
+          key: 'Tonga',
+          title: 'Tonga',
+        },
+      ],
+      rows: [
+        {
+          categoryId: 'medical center',
+          dataElement: 'hospital',
+          Laos: 3,
+          Tonga: 0,
+        },
+        {
+          categoryId: 'medical center',
+          dataElement: 'clinic',
+          Laos: 4,
+          Tonga: 9,
+        },
+        {
+          categoryId: 'others',
+          dataElement: 'park',
+          Laos: 2,
+          Tonga: 0,
+        },
+        {
+          categoryId: 'others',
+          dataElement: 'library',
+          Laos: 0,
+          Tonga: 5,
+        },
+        {
+          category: 'medical center',
+        },
+        {
+          category: 'others',
+        },
+      ],
+    };
+
+    it('can perform without columns', () => {
+      const output = buildOutput({
+        matrix: {
+          categoryField: 'InfrastructureType',
+          rowField: 'FacilityType',
+        },
+      });
+      expect(output(MULTIPLE_TRANSFORMED_DATA)).toEqual(expectedData);
+    });
+
+    it("throw error when columns is not '*' or string array", () => {
+      const output = buildOutput({
+        matrix: {
+          categoryField: 'InfrastructureType',
+          rowField: 'FacilityType',
+          columns: 1,
+        },
+      });
+      expect(() => {
+        output(MULTIPLE_TRANSFORMED_DATA);
+      }).toThrow();
+    });
+
+    it('can perform with categoryField', () => {
+      const output = buildOutput({
+        matrix: {
+          categoryField: 'InfrastructureType',
+          rowField: 'FacilityType',
+          columns: '*',
+        },
+      });
+      expect(output(MULTIPLE_TRANSFORMED_DATA)).toEqual(expectedData);
+    });
```

Comment:
Might just wanna add a test to check that `it('can return only specified columns')` where something like `columns: ['Laos']` is in the config 👍

Reply:
@rohan-bes Sounds good! Do we also want to eliminate other unspecified column keys in rows? We have Laos and Tonga in this expectedData, currently we will shows them all. Perhaps the best practice is to eliminate other unspecified rows data for Tonga?

Reply:
>Do we also want to eliminate other unspecified column keys in rows?
Yeah, I think that would be best. It's a bit unexpected to have a field in a row that isn't specified in the columns 👍